sdk/metric/reader.go

```diff
@@ -139,3 +140,47 @@ func (t temporalitySelectorOption) applyPeriodic(prc periodicReaderConfig) perio
 func defaultTemporalitySelector(InstrumentKind) Temporality {
 	return CumulativeTemporality
 }
+
+// WithAggregation sets the default aggregation a reader will use for an
+// instrument based on the returned value from the selector. If this option is
+// not used, the reader will use the default selector which uses the following
+// selection mapping: Counter ⇨ Sum, Asynchronous Counter ⇨ Sum, UpDownCounter
+// ⇨ Sum, Asynchronous UpDownCounter ⇨ Sum, Asynchronous Gauge ⇨ LastValue,
+// Histogram ⇨ ExplicitBucketHistogram.
+func WithAggregation(selector func(InstrumentKind) aggregation.Aggregation) ReaderOption {
```

Comment:
It is odd to me that this doesn't take an aggregation.Aggregation.  How do I make a "selector"?

Reply:
Would it help if renamed to `WithAggregationSeletor`?

For answer completeness, you will need to provide a function similar to the [defaultAggregationSelector](https://github.com/open-telemetry/opentelemetry-go/pull/2958/files#diff-8aaf470136f6caba7b53dd80f9d293cb49d1b46b57061e4a4b69d6642b544f5bR172). Would it help if we made `func(InstrumentKind) aggregation.Aggregation` it's own type and document it?

Reply:
The definitions of [Aggregations](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#default-aggregation) in the spec look more like mapping functions than like values to me, but from instrument kind to concrete types.  I had assumed passing in `WithAggregation(aggregation.Default)` here is what would give me the defaultAggregationSelector under the hood, and find it surprising that it isn't the case.

I think your suggestions would definitely help.  I was mostly just confused trying to reconcile the spec to this.

Reply:
My read of the specification is there are two places aggregations can be specified: the reader, and the view. At the reader they are specified as a mapping function, every instrument type is mapped to an aggregation.

> The aggregation (optional) to be used. If not provided, the SDK MUST apply a default aggregation configurable on the basis of instrument kind according to the MetricReader instance.

and

> If the MeterProvider has no View registered, take the Instrument and apply the default Aggregation on the basis of instrument kind according to the MetricReader instance's aggregation property.

But, at the view it is a singular aggregation:

> An Aggregation, as configured via the View, informs the SDK on the ways and means to compute Aggregated Metrics from incoming Instrument Measurements.

Instrument being singular here. I read the `Deafault` option as a way for a view to explicitly go around a readers aggregation mapping and say something like "even if the reader says use last-value for this counter, use the default sum instead" (similarly for all the other instrument types.

I'll update the `WithAggregation` reader function name, and update `WithTemporality` to match (it also is accepts a selector).

Reply:
@dashpole please let me know if f719d40fbc4236eba9a650f1b6d6c5a8ca223a70 addresses this. I'll update `WithTemporality` to match if so.

Reply:
Is does, thank you :)